### PR TITLE
Add buffering to ChannelTransport

### LINKED ISSUE #2125: HTTP/2 stack creates excessive latency and throughput overhead
Issue Type: *Bug report*

Linkerd introduces a significant amount of overhead in terms of both throughput and latency for gRPC services versus two processes communicating directly. 

Here is a max qps [strest-grpc](http://github.com/buoyantio/strest-grpc) run for each scenario: 

#### Direct

```bash
> ./strest-grpc server --address "127.0.0.1:9999"
> ./strest-grpc client --address "127.0.0.1:9999" --totalRequests 100000 --streams 100 
{
  "good": 100000,
  "bad": 0,
  "bytes": 0,
  "latency": {
    "p50": 3,
    "p75": 4,
    "p90": 5,
    "p95": 5,
    "p99": 8,
    "p999": 19
  },
  "jitter": {
    "p50": 0,
    "p75": 0,
    "p90": 0,
    "p95": 0,
    "p99": 0,
    "p999": 0
  }
}
```

#### Via Linkerd

```bash
> ./strest-grpc client --address "127.0.0.1:4143" --totalRequests 100000 --streams 100 
{
  "good": 100000,
  "bad": 0,
  "bytes": 0,
  "latency": {
    "p50": 46,
    "p75": 53,
    "p90": 76,
    "p95": 122,
    "p99": 229,
    "p999": 412
  },
  "jitter": {
    "p50": 0,
    "p75": 0,
    "p90": 0,
    "p95": 0,
    "p99": 0,
    "p999": 0
  }
}
```

#### Configuration:
```yaml
admin:
  port: 9990
  ip: 0.0.0.0

routers: 
  - label: h2-in
    protocol: h2
    experimental: true
    client:
      initialStreamWindowBytes: 1048576
      failureAccrual:
        kind: none
    servers:
      - port: 4143
        ip: 0.0.0.0
        maxConcurrentStreamsPerConnection: 2147483647
        initialStreamWindowBytes: 1048576
    dtab: |
       /svc/* => /$/inet/127.0.0.1/9999;
    identifier:
      kind: io.l5d.header.path
      segments: 1
```

In addition to using strest-grpc, I also wrote a custom but crude benchmarking tool of my own, [echobench](http://github.com/zackangelo/echobench),  so that I could experiment with gRPC channel settings and socket options.

The custom tool also reports a significant amount of overhead when using linkerd: 

#### Direct
```
=== summary ===
threads: 10 (5000 reqs per thread)
requests: 50000
throughput: 2922.804486159504/s
errors: 0
latency:
- min: 1ms
- max: 17ms
- median: 2.0ms
- avg: 2.8737975505112257ms
- p95: 7.0ms
- p99: 11.0ms
```

#### Via Linkerd
```
=== summary ===
threads: 10 (5000 reqs per thread)
requests: 50000
throughput: 629.5124804747321/s
errors: 0
latency:
- min: 6ms
- max: 34ms
- median: 13.0ms
- avg: 13.981156425923315ms
- p95: 20.0ms
- p99: 27.0ms
```

Checking the `metrics.json` endpoint after a test run shows an interesting discrepancy between request and stream latencies. Stream latencies appear to be much higher: 

#### Request latencies
```
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.count": 53932,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.max": 54,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.min": 0,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.p50": 6,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.p90": 16,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.p95": 22,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.p99": 31,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.p9990": 41,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.p9999": 50,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.sum": 390436,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/request_latency_ms.avg": 7.239412593636431,
```

#### Stream latencies
```
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.count": 53945,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.max": 139,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.min": 10,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.p50": 46,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.p90": 59,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.p95": 63,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.p99": 79,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.p9990": 104,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.p9999": 135,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.sum": 2538980,
  "rt/h2-in/client/$/inet/127.0.0.1/9999/service/svc/echo.EchoService/response/stream/stream_duration_ms.avg": 47.06608582815831,
```

Another interesting peculiarity is there is only ever one stream indicated as open back to the test service, despite several threads sending requests simultaneously: 

```
"rt/h2-in/client/$/inet/127.0.0.1/9999/stream/open_streams": 1
```

After discovering the overhead I attempted several configuration and code changes to alleviate this issue, including: 

* Manually defining max concurrent stream values to maximum values on both ends of the connection
* Manually expanding HTTP/2 flow control window sizes to maximum allowable values on both ends of the connection
* Setting the retry buffer sizes to 0
* Removing the `ClassifiedRetries` module from the finagle client stack entirely in code
* Manually expanding the maximum frame size to maximum allowable values
* Removing the [DelayedReleaseService](https://github.com/linkerd/linkerd/blob/master/finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/DelayedReleaseService.scala)  from the finagle client stack in code

None of these changes made a noticeable impact on the latency and throughput. 

I took several tcpdumps to examine the behavior of the service communication under both scenarios.  A big difference number of captured packets immediately jumps out: 2966 for direct and 22542 for linkerd. 

Looking at the tcpdumps in Wireshark confirms why: when communicating directly the client is able to multiplex HTTP/2 frames from ~120 streams into a single TCP packet. Linkerd, by comparison, creates a TCP packet for each HTTP/2 stream frame.

#### Linkerd packets 
image

#### Direct packets
image

The discrepancy in latency and throughput could be attributed to the additional syscall and packetization overhead required to forward HTTP/2 traffic in this way.

I found the discrepancy surprising given that both linkerd and grpc-java are based on the netty4 HTTP/2 stack primitives. It turns out that when the grpc-java stack was being developed, they identified a problem with netty4’s H2 stack flushing too frequently. 

They identified and fixed the issue in these places: 

* [HTTP2 remote flow controller flushing is too eager · Issue #3688 · netty/netty · GitHub](https://github.com/netty/netty/issues/3688)
* [HTTP/2 shouldn’t flush automatically · Issue #3670 · netty/netty · GitHub](https://github.com/netty/netty/issues/3670)
* [Disable flushing on frame write in flow-controller by louiscryan · Pull Request #3691 · netty/netty · GitHub](https://github.com/netty/netty/pull/3691)
* [Reduce number of flushes · Issue #305 · grpc/grpc-java · GitHub](https://github.com/grpc/grpc-java/issues/305)
* [Implement writes to the channel using a dedicated write queue by louiscryan · Pull Request #431 · grpc/grpc-java · GitHub](https://github.com/grpc/grpc-java/pull/431)

The benchmarks listed in the write queue PR indicate this change unlocked a significant jump in throughput, especially in the case of many streams (16806.718 ops/s versus 55975.008 ops/s for 1000 streams). 

I don’t have an easy way to confirm how much a write queue will reduce the performance overhead I’m observing. I’m open to suggestions for a quick and dirty way to get more confidence in this as a potential solution.

I did try enabling Nagle’s (removing the `TCP_NODELAY` option) from the client socket in linkerd and after a few subsequent test runs was able to get the throughput to increase in a single threaded use-case.

### PR DESCRIPTION
**Pull request by adleong**
Fixes #2125 

When Linkerd writes a frame to the Netty4 ChannelTransport, it writes the frame and then immediately flushing it, causing a system call for each frame.  In cases where many small frames are being written, this can cause a high amount of overhead and limit throughput.  See #2125 for details.

By buffering writes and scheduling a job on the channel event loop to flush them, we can consolidate many writes into a single flush.  This approach was inspired by similar work done in grpc-java: https://github.com/grpc/grpc-java/pull/431

In my testing, this leads to a 15-30% improvement in throughput.

```
# MASTER

alex@ares:~$ time ./strest-grpc client --address "127.0.0.1:4143" --totalRequests 200000 --streams 100
2018-10-01T12:51:30-07:00    0.0B   7079/0 10s L:   0 [330 1126 ] 1242 J:   0   0
2018-10-01T12:51:40-07:00    0.0B  31917/0 10s L:   0 [ 62  88 ]  126 J:   0   0
2018-10-01T12:51:50-07:00    0.0B  42177/0 10s L:   0 [ 38  53 ]   90 J:   0   0
2018-10-01T12:52:00-07:00    0.0B  40807/0 10s L:   0 [ 40  56 ]   84 J:   0   0
2018-10-01T12:52:10-07:00    0.0B  40823/0 10s L:   0 [ 41  57 ]  101 J:   0   0
2018-10-01T12:52:20-07:00    0.0B  37197/0 10s L:   0 [ 44  55 ]   87 J:   0   0
{
  "good": 200000,
  "bad": 0,
  "bytes": 0,
  "latency": {
    "50": 22,
    "95": 57,
    "99": 148,
    "999": 1242
  },
  "jitter": {
    "50": 0,
    "95": 0,
    "99": 0,
    "999": 0
  }
}

real    0m59.170s
user    0m19.844s
sys     0m8.816s

# WITH BUFFERING

$ time ./strest-grpc client --address "127.0.0.1:4143" --totalRequests 200000 --streams 100
2018-10-01T16:56:40-07:00    0.0B   8738/0 10s L:   0 [291 1098 ] 1149 J:   0   0
2018-10-01T16:56:50-07:00    0.0B  46768/0 10s L:   0 [ 41  67 ]  140 J:   0   0
2018-10-01T16:57:00-07:00    0.0B  50951/0 10s L:   0 [ 37  50 ]   85 J:   0   0
2018-10-01T16:57:10-07:00    0.0B  50365/0 10s L:   0 [ 37  52 ]   79 J:   0   0
2018-10-01T16:57:20-07:00    0.0B  43178/0 10s L:   0 [ 38  56 ]   90 J:   0   0
{
  "good": 200000,
  "bad": 0,
  "bytes": 0,
  "latency": {
    "50": 17,
    "95": 50,
    "99": 136,
    "999": 1149
  },
  "jitter": {
    "50": 0,
    "95": 0,
    "99": 0,
    "999": 0
  }
}

real    0m48.585s
user    0m19.020s
sys     0m3.656s
```

### CODE REVIEW

**Comment by adleong** [comment]
cc @zackangelo

**Comment by zackangelo** [comment]
@adleong do you know if this resulted in a latency improvement versus connecting directly?

**Comment by adleong** [comment]
This change reduces the latency overhead added by Linkerd but it's still substantial (50ms p90 in my above test).  You can see the full data in the slow_cooker output above.

File: finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  // Always flush after this many messages.
+  private[this] val MaxFlushSize = 128
+  private[this] val flushScheduled = new AtomicBoolean(false)
+  private[this] val writeQueue = new LinkedBlockingQueue[WriteItem]()
+  private[this] val writeChunk = new util.ArrayDeque[WriteItem](MaxFlushSize)
+
+  override def write(msg: Any): Future[Unit] = {
+    val p = new Promise[Unit]()
+    writeQueue.add(WriteItem(msg, p))
+    scheduleFlush()
```
Comment (dadjeibaah) [inline finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala]: Does `scheduleFlush()` need to happen when `writeQueue` has reached `MaxFlushSize` of 128? Does this need to be guarded against a condition?

File: finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      while (writeChunk.size > 0) {
+        val item = writeChunk.poll()
+        val f = toFuture(ch.write(item.msg))
+        item.done.become(f)
+      }
+      flushed = true
+      ch.flush()
+    }
+    // Always flush at least once.
+    if (!flushed) {
```
Comment (dadjeibaah) [inline finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala]: Curious as to why we need to flush at least once? Is this true in the case where we do not have any messages to write?

File: finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  ch: nettyChan.Channel,
+  readQueue: AsyncQueue[Any] = new AsyncQueue[Any],
+  omitStackTraceOnInactive: Boolean = false
+) extends ChannelTransport(ch, readQueue, omitStackTraceOnInactive) {
+
+  // Satisfy the done promise when the write completes.
+  private case class WriteItem(msg: Any, done: Promise[Unit])
+
+  // Always flush after this many messages.
+  private[this] val MaxFlushSize = 128
```
[the remark this answers is not in the document]
Comment (adleong) [reply finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala]: Cargo culted from grpc-java: https://github.com/grpc/grpc-java/pull/431/files#diff-7f048858dab93d58f2bcac583626abddR49

This is mostly just a safety net to ensure that we don't buffer up arbitrarily large writes without flushing.  In most cases I would expect us to flush before hitting this limit.

File: finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala (hunk 1 above)
Comment (adleong) [reply finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala]: I don't think so.  All writes will add the msg to the writeQueue and schedule a flush if one has not already been scheduled.  During the scheduled flush, we will drain off the writeQueue 128 items at a time and write them out as a group.

File: finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala (hunk 2 above)
Comment (adleong) [reply finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala]: Cargo culted from java-grpc: https://github.com/grpc/grpc-java/pull/431/files#diff-7f048858dab93d58f2bcac583626abddR132

I'm not actually sure, but I suspect this is just to guarantee that we can't have writes that get stuck in the Channel buffer forever and never get flushed.

**Comment by adleong** [comment]
@zackangelo if you get a chance, I'd love to hear if this change yields the above performance benefits in your environment.

File: finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (olix0r) [reply finagle/h2/src/main/scala/com/twitter/finagle/netty4/transport/buoyant/BufferingChannelTransport.scala]: worth including that link for next time we have to read this?

**Pull request merged**